hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/source/FileIndex.java

```diff
@@ -60,23 +61,36 @@ public class FileIndex {
 
   private final Path path;
   private final RowType rowType;
+  private final boolean tableExists;
   private final HoodieMetadataConfig metadataConfig;
+  private final PartitionPruner partitionPruner;
   private final boolean dataSkippingEnabled;
+  private final DataPruner dataPruner;
   private List<String> partitionPaths;      // cache of partition paths
-  private List<ResolvedExpression> filters; // push down filters
-  private final boolean tableExists;
-  private DataPruner dataPruner;
 
-  private FileIndex(Path path, Configuration conf, RowType rowType) {
+  private FileIndex(Path path, Configuration conf, RowType rowType, DataPruner dataPruner, PartitionPruner partitionPruner) {
     this.path = path;
     this.rowType = rowType;
+    this.tableExists = StreamerUtil.tableExists(path.toString(), HadoopConfigurations.getHadoopConf(conf));
     this.metadataConfig = metadataConfig(conf);
     this.dataSkippingEnabled = conf.getBoolean(FlinkOptions.READ_DATA_SKIPPING_ENABLED);
-    this.tableExists = StreamerUtil.tableExists(path.toString(), HadoopConfigurations.getHadoopConf(conf));
+    // NOTE: Data Skipping is only effective when it references columns that are indexed w/in
+    //       the Column Stats Index (CSI). Following cases could not be effectively handled by Data Skipping:
+    //          - Expressions on top-level column's fields (ie, for ex filters like "struct.field > 0", since
+    //          CSI only contains stats for top-level columns, in this case for "struct")
+    //          - Any expression not directly referencing top-level column (for ex, sub-queries, since there's
+    //          nothing CSI in particular could be applied for)
+    if (!metadataConfig.enabled() || !dataSkippingEnabled) {
+      validateConfig();
+      this.dataPruner = null;
+    } else {
+      this.dataPruner = dataPruner;
+    }
+    this.partitionPruner = partitionPruner;
```

Comment:
can we move the instantiation of `dataPruner` into a separate method named like `initDataPruner` ?